### PR TITLE
Add multi field union iterator

### PR DESCRIPTION
**Pull request by xichen2020**
cc @black-adder @notbdu 

This PR adds support for multi-field union iterator, and refactors the multi-field intersect iterator to reuse the `inAnyDocIDSetIterator` and `inAllDocIDSetIterator` to reduce repeated logic.

### CODE REVIEW

File: index/field/multi_field_intersect_iterator.go [hunk 1]
```diff
@@ -2,91 +2,56 @@ package field
...
-		}
-		if !it.iters[minIdx].Next() {
-			it.done = true
-			it.err = it.iters[minIdx].Err()
-			return false
-		}
-		it.currDocIDs[minIdx] = it.iters[minIdx].DocID()
-	}
-}
+func (it *MultiFieldIntersectIterator) Next() bool { return it.intersectIter.Next() }
```
Comment (notbdu) [inline index/field/multi_field_intersect_iterator.go]: Looking at the implementation of `InAnyDocIDSetIterator`, it looks like it does not perform the intersect operation. As in, it only advances each underlying base iterator forward w/o ensuring that all DocIDs line up.

It looks like it's used below in the `MultiFieldUnionIterator` which makes sense. But in that case, aren't we missing the intersect operation here? Maybe I'm missing something though.

Comment (xichen2020) [reply index/field/multi_field_intersect_iterator.go]: Note that this uses `InAllDocIDSetIterator` which advances all doc ID iterators in sync.

Comment (notbdu) [reply index/field/multi_field_intersect_iterator.go]: Oh woops disregard my comment then.

**Comment by notbdu** [comment]
LGTM.

File: index/field/multi_field_union_iterator.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+func (it *MultiFieldUnionIterator) DocID() int32 { return it.unionIter.DocID() }
+
+// Values returns the current list of field values, which remains valid until the
+// next iteration. If the caller needs to retain a valid refence to the value array
+// after `Next` is called again, the caller needs to make a copy of the value array.
+// Caller should check the boolean array to see whether the corresponding slot in the
+// value array is valid.
+func (it *MultiFieldUnionIterator) Values() ([]field.ValueUnion, []bool) {
+	for i := 0; i < len(it.iters); i++ {
+		if it.unionIter.IsDoneAt(i) || it.unionIter.DocID() != it.iters[i].DocID() {
```
Comment (black-adder) [inline index/field/multi_field_union_iterator.go]: can this whole check be encapsulated inside `InAnyDocIDSetIterator`? `it.unionIter.IsValidAt` or something?

Comment (xichen2020) [reply index/field/multi_field_union_iterator.go]: Fair enough, will move

**Pull request merged**